cudamapper/src/chainer_utils.cu

```diff
@@ -0,0 +1,221 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "chainer_utils.cuh"
+
+#include <cub/cub.cuh>
+#include <claraparabricks/genomeworks/utils/cudautils.hpp>
+
+namespace claraparabricks
+{
+
+namespace genomeworks
+{
+
+namespace cudamapper
+{
+namespace chainerutils
+{
+
+__host__ __device__ Overlap create_simple_overlap(const Anchor& start, const Anchor& end, const int32_t num_anchors)
+{
+    Overlap overlap;
+    overlap.num_residues_ = num_anchors;
+
+    overlap.query_read_id_  = start.query_read_id_;
+    overlap.target_read_id_ = start.target_read_id_;
+    assert(start.query_read_id_ == end.query_read_id_ && start.target_read_id_ == end.target_read_id_);
+
+    overlap.query_start_position_in_read_ = min(start.query_position_in_read_, end.query_position_in_read_);
+    overlap.query_end_position_in_read_   = max(start.query_position_in_read_, end.query_position_in_read_);
+    const bool is_negative_strand         = end.target_position_in_read_ < start.target_position_in_read_;
+    if (is_negative_strand)
+    {
+        overlap.relative_strand                = RelativeStrand::Reverse;
+        overlap.target_start_position_in_read_ = end.target_position_in_read_;
+        overlap.target_end_position_in_read_   = start.target_position_in_read_;
+    }
+    else
+    {
+        overlap.relative_strand                = RelativeStrand::Forward;
+        overlap.target_start_position_in_read_ = start.target_position_in_read_;
+        overlap.target_end_position_in_read_   = end.target_position_in_read_;
+    }
+    return overlap;
+}
+
+void allocate_anchor_chains(device_buffer<Overlap>& overlaps,
+                            device_buffer<int32_t>& unrolled_anchor_chains,
+                            device_buffer<int32_t>& anchor_chain_starts,
+                            int32_t num_overlaps,
+                            int32_t& num_total_anchors,
+                            DefaultDeviceAllocator& _allocator,
+                            cudaStream_t& _cuda_stream)
+{
+    // sum the number of chains across all overlaps
+    device_buffer<char> d_temp_buf(_allocator, _cuda_stream);
+    void* d_temp_storage           = nullptr;
+    std::size_t temp_storage_bytes = 0;
+    OverlapToNumResiduesOp overlap_residue_count_op;
+    cub::TransformInputIterator<int32_t, OverlapToNumResiduesOp, Overlap*> d_residue_counts(overlaps.data(), overlap_residue_count_op);
+
+    device_buffer<int32_t> d_num_total_anchors(1, _allocator, _cuda_stream);
+
+    cub::DeviceReduce::Sum(d_temp_storage,
+                           temp_storage_bytes,
+                           d_residue_counts,
+                           d_num_total_anchors.data(),
+                           num_overlaps,
+                           _cuda_stream);
+
+    d_temp_buf.clear_and_resize(temp_storage_bytes);
+    d_temp_storage = d_temp_buf.data();
+
+    cub::DeviceReduce::Sum(d_temp_storage,
```

Comment:
If I understand correctly you are just doing a reduction here? Could this be replaced by `thrust::reduce`? In that case that would result in less code.
For the way to specify allocator and stream to thrust see https://github.com/clara-parabricks/GenomeWorks/blob/34d1d5231fd6385adf146809fd1c36cec4f1a841/common/base/include/claraparabricks/genomeworks/utils/cudasort.cuh#L171

Reply:
Do thrust/CUB now have the same performance? The CUB docs seem to indicate better performance than thrust. A quick Google search seems to provide little evidence either way.

Reply:
It depends. Thrust is largely built on top on CUB so in general Thrust should not be faster than CUB (not sure if there are exception to this rule).
Going in the opposite direction, CUB is more fine-grained so it's sometimes possible to get better performance using some problem-specific properties. For example in `sort_by_two_keys()` after first sort we only move the indices instead of the whole array and we use CUB's API which allows to specify the number of bytes that have to be sorted as we know that not all keys are possible: https://github.com/clara-parabricks/GenomeWorks/blob/bc855b37a62620a8f32388bffc426a334dbbaded/common/base/include/claraparabricks/genomeworks/utils/cudasort.cuh#L139
As a rule of thumb I would suggest using Thrust as it is more readable and only switching to CUB if there are such problem-specific properties or if profiling shows that that is a bottleneck so it would be worth giving CUB a try

Reply:
Swapped to thrust in 6244636 .